ports/esp32/machine_can.c

```diff
@@ -0,0 +1,673 @@
+        mask = ((mask & 0x7FF) << 5);
+        preserve_mask = 0xFFFF << (bank == 0 ? 16 : 0);
+        if ((self->config->filter.acceptance_mask & preserve_mask) == (0xFFFF << (bank == 0 ? 16 : 0))){
+            // Other filter accepts all; it will replaced duplicating current filter
+            addr = addr | (addr << 16);
+            mask = mask | (mask << 16);
+            preserve_mask = 0;
+        }else{
+            addr = addr << (bank == 1 ? 16 : 0);
+            mask = mask << (bank == 1 ? 16 : 0);
+        }
+    }
+    self->config->filter.acceptance_code &= preserve_mask;
+    self->config->filter.acceptance_code |= addr;
+    self->config->filter.acceptance_mask &= preserve_mask;
+    self->config->filter.acceptance_mask |= mask;
+}
+
+// Force a software restart of the controller, to allow transmission after a bus error
+STATIC mp_obj_t machine_hw_can_restart(mp_obj_t self_in){
+    uint32_t status = can_initiate_recovery();
+    if (status != ESP_OK){
+        mp_raise_OSError(-status);
+    }
+    mp_hal_delay_ms(200);
+    status = can_start();
+    if (status != ESP_OK){
+        mp_raise_OSError(-status);
+    }
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_hw_can_restart_obj, machine_hw_can_restart);
+
+// any() - return `True` if any message waiting, else `False`
+STATIC mp_obj_t machine_hw_can_any(mp_obj_t self_in){
+    can_status_info_t status = _machine_hw_can_get_status();
+    return mp_obj_new_bool((status.msgs_to_rx) > 0);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_hw_can_any_obj, machine_hw_can_any);
+
+// Get the state of the controller
+STATIC mp_obj_t machine_hw_can_state(mp_obj_t self_in){
+    can_status_info_t status = _machine_hw_can_get_status();
+    return mp_obj_new_int(status.state);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_hw_can_state_obj, machine_hw_can_state);
+
+// Get info about error states and TX/RX buffers
+STATIC mp_obj_t machine_hw_can_info(size_t n_args, const mp_obj_t *args){
+    machine_can_obj_t *self = MP_OBJ_TO_PTR(args[0]);
+    mp_obj_list_t *list;
+    if (n_args == 1){
+        list = MP_OBJ_TO_PTR(mp_obj_new_list(8, NULL));
+    }else{
+        if (!mp_obj_is_type(args[1], &mp_type_list)){
+            mp_raise_TypeError(NULL);
+        }
+        list = MP_OBJ_TO_PTR(args[1]);
+        if (list->len < 8){
+            mp_raise_ValueError(NULL);
+        }
+    }
+    can_status_info_t status = _machine_hw_can_get_status();
+    list->items[0] = MP_OBJ_NEW_SMALL_INT(status.tx_error_counter);
+    list->items[1] = MP_OBJ_NEW_SMALL_INT(status.rx_error_counter);
+    list->items[2] = MP_OBJ_NEW_SMALL_INT(self->num_error_warning);
+    list->items[3] = MP_OBJ_NEW_SMALL_INT(self->num_error_passive);
+    list->items[4] = MP_OBJ_NEW_SMALL_INT(self->num_bus_off);
+    list->items[5] = MP_OBJ_NEW_SMALL_INT(status.msgs_to_tx);
+    list->items[6] = MP_OBJ_NEW_SMALL_INT(status.msgs_to_rx);
+    list->items[7] = mp_const_none;
```

Comment:
What's this for? I see the documentation just says "Reserved"

Reply:
Current ESP-IDF implementation for CAN device do not foresee events management. So in order to have a cross-compatibility with STM32 I left the same array order information, but for those information not available I indicated RESERVED in the documentation. Maybe for the next implementation step, I'll introduce some free RTOS task to monitor the status of device and give the information to the mp layer.

Reply:
Ah yes, sorry I missed this in the STM32 version. Thanks